File: levenshtein/julia/levenshtein.jl
module Levenshtein

export levenshtein_distance, distances

# Calculates the Levenshtein distance between two strings using the Wagner-Fischer algorithm.
# Space Complexity: O(min(m,n)) – allocates two arrays per call.
# Time Complexity: O(m*n) where m and n are the lengths of the input strings.
function levenshtein_distance(str1::String, str2::String)::Int
    # Early termination checks
    if str1 == str2
        return 0
    end
    s1 = codeunits(str1)
    s2 = codeunits(str2)
    if isempty(s1)
        return length(s2)
    elseif isempty(s2)
        return length(s1)
    end

    # Make s1 the shorter string for space optimization
    if length(s1) > length(s2)
        s1, s2 = s2, s1
    end

    m, n = length(s1), length(s2)

    prev_row = Vector{Int32}(undef, m + 1)
    curr_row = Vector{Int32}(undef, m + 1)

    # Initialize first row
    for i in 0:m
        @inbounds prev_row[i + 1] = Int32(i)
    end

    # Main computation loop
    @inbounds for j in 1:n
        curr_row[1] = Int32(j)
        for i in 1:m
            cost = s1[i] == s2[j] ? 0 : 1
            @inbounds curr_row[i + 1] = min(
                prev_row[i + 1] + 1,    # deletion
                curr_row[i] + 1,        # insertion
                prev_row[i] + cost      # substitution
            )
        end
        # Swap rows (we simply swap the array references)
        prev_row, curr_row = curr_row, prev_row
    end

    return @inbounds prev_row[m + 1]
end

# distances : Vector{String} -> Vector{Int}
#
# Computes the Levenshtein distance for every unique unordered pair of strings.
# That is, for each pair (A, B) with A ≠ B the distance is computed only once.
function distances(strings::Vector{String})
    n = length(strings)
    count = n*(n-1) ÷ 2
    dists = Vector{Int}(undef, count)
    idx = 1
    for i in 1:(n-1)
        for j in (i+1):n
            dists[idx] = levenshtein_distance(strings[i], strings[j])
            idx += 1
        end
    end
    return dists
end

end # module Levenshtein

File: levenshtein/swift/levenshtein.swift
/// Calculates the Levenshtein distance between two strings using Wagner-Fischer algorithm
/// Space Complexity: O(min(m,n)) - only uses two arrays instead of full matrix
/// Time Complexity: O(m*n) where m and n are the lengths of the input strings
func levenshteinDistance(_ s1: String, _ s2: String) -> Int {
    // Early termination checks
    if s1 == s2 { return 0 }
    if s1.isEmpty { return s2.count }
    if s2.isEmpty { return s1.count }

    // Make s1 the shorter string for space optimization
    let (str1, str2) = s1.count > s2.count ? (s2, s1) : (s1, s2)

    // Convert strings to arrays for O(1) access
    let arr1 = str1.utf8CString.dropLast()
    let arr2 = str2.utf8CString.dropLast()

    let m = arr1.count
    let n = arr2.count

    // Use two arrays instead of full matrix for space optimization
    var prevRow = [Int](0...m)
    var currRow = [Int](repeating: 0, count: m + 1)

    // Main computation loop
    for j in 1...n {
        currRow[0] = j

        for i in 1...m {
            let cost = arr1[i-1] == arr2[j-1] ? 0 : 1

            // Calculate minimum of three operations
            currRow[i] = min(
                prevRow[i] + 1,        // deletion
                currRow[i-1] + 1,      // insertion
                prevRow[i-1] + cost    // substitution
            )
        }

        // Swap rows
        (prevRow, currRow) = (currRow, prevRow)
    }

    return prevRow[m]
}

// New function to compute distances for unique string pairs
public func distances(_ words: [String]) -> [Int] {
    var result: [Int] = []
    for i in 0..<words.count {
        for j in (i + 1)..<words.count {
            result.append(levenshteinDistance(words[i], words[j]))
        }
    }
    return result
}
